File: backend/fretquiz/src/main/java/kdp/fretquiz/websocket/Request.java
package kdp.fretquiz.websocket;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import kdp.fretquiz.game.Guess;
import kdp.fretquiz.theory.Accidental;

/**
 * A Request is a message originating from the client (the browser).
 * Each Request has a Request.Type and, optionally, some associated data.
 */
public interface Request
{
    Type type();

    enum Type
    {
        LOGIN,
        LOGOUT,
        GET_GAME_IDS,
        CREATE_GAME,
        JOIN_GAME,
        START_GAME,
        PLAYER_GUESS,
        NEXT_ROUND,
        TOGGLE_STRING,
        TOGGLE_ACCIDENTAL,
        SET_ROUND_COUNT,
        START_COUNTDOWN,
        START_ROUND_COUNTDOWN
    }

    /**
     * Request.Default is an record that only contains the incoming Request's type.
     * Each incoming request is cast to Request.Default first to determine the Request type.
     * Then it is then cast to the proper Request type so it can be used further.
     */
    @JsonIgnoreProperties(ignoreUnknown = true)
    record Default(Type type) implements Request
    {
    }

    record Login(Type type, String name) implements Request
    {
        public Login(String name)
        {
            this(Type.LOGIN, name);
        }
    }

    record Logout(Type type) implements Request
    {
        public Logout()
        {
            this(Type.LOGOUT);
        }
    }

    record GetGameIds(Type type) implements Request
    {
        public GetGameIds()
        {
            this(Type.GET_GAME_IDS);
        }
    }

    record CreateGame(Type type) implements Request
    {
        public CreateGame()
        {
            this(Type.CREATE_GAME);
        }
    }

    record JoinGame(Type type,
                    String gameId,
                    String userId) implements Request
    {
        public JoinGame(String gameId, String userId)
        {
            this(Type.JOIN_GAME, gameId, userId);
        }
    }

    record StartGame(Type type, String gameId)
    {
        public StartGame(String gameId)
        {
            this(Type.START_GAME, gameId);
        }
    }

    record PlayerGuess(Type type, Guess.ClientGuess clientGuess) implements Request
    {
        public PlayerGuess(Guess.ClientGuess clientGuess)
        {
            this(Type.PLAYER_GUESS, clientGuess);
        }
    }

    record NextRound(Type type, String gameId, String userId) implements Request
    {
        public NextRound(String gameId, String userId)
        {
            this(Type.NEXT_ROUND, gameId, userId);
        }
    }

    record ToggleString(Type type, String gameId, int string) implements Request
    {
        public ToggleString(String gameId, int string)
        {
            this(Type.TOGGLE_STRING, gameId, string);
        }
    }

    record ToggleAccidental(Type type, String gameId, Accidental accidental) implements Request
    {
        public ToggleAccidental(String gameId, Accidental accidental)
        {
            this(Type.TOGGLE_ACCIDENTAL, gameId, accidental);
        }
    }

    record SetRoundCount(Type type, String gameId, int roundCount) implements Request
    {
        public SetRoundCount(String gameId, int roundCount)
        {
            this(Type.SET_ROUND_COUNT, gameId, roundCount);
        }
    }

    record StartCountdown(Type type, String gameId) implements Request
    {
        public StartCountdown(String gameId)
        {
            this(Type.START_COUNTDOWN, gameId);
        }
    }

    record StartRoundCountdown(Type type, String gameId) implements Request
    {
        public StartRoundCountdown(String gameId)
        {
            this(Type.START_ROUND_COUNTDOWN, gameId);
        }
    }
}

File: frontend/fretquiz/src/websocket/request.ts
import { ClientGuess, Accidental } from '../game/types';

export type RequestType =
  'LOGIN'
  | 'LOGOUT'
  | 'GET_GAME_IDS'
  | 'CREATE_GAME'
  | 'JOIN_GAME'
  | 'START_GAME'
  | 'PLAYER_GUESS'
  | 'NEXT_ROUND'
  | 'TOGGLE_STRING'
  | 'TOGGLE_ACCIDENTAL'
  | 'SET_ROUND_COUNT'
  | 'START_COUNTDOWN'
  | 'START_ROUND_COUNTDOWN';

export interface Request {
  type: RequestType
}

// user requests

export interface Login extends Request {
  type: 'LOGIN',
  name: string
}

export function login(name: string): Login {
  return { type: 'LOGIN', name };
}

export interface Logout extends Request {
  type: 'LOGOUT'
}

export function logout(): Logout {
  return { type: 'LOGOUT' }
}

// game requests

export interface GetGameIds extends Request {
  type: 'GET_GAME_IDS'
}

export function getGameIds(): GetGameIds {
  return { type: 'GET_GAME_IDS' }
}

export interface CreateGame extends Request {
  type: 'CREATE_GAME'
}

export function createGame(): CreateGame {
  return { type: 'CREATE_GAME' }
}

export interface JoinGame extends Request {
  type: 'JOIN_GAME',
  gameId: string,
  userId: string
}

export function joinGame(gameId: string, userId: string): JoinGame {
  return { type: 'JOIN_GAME', gameId, userId: userId }
}

export interface StartGame extends Request {
  type: 'START_GAME',
  gameId: string
}

export function startGame(gameId: string): StartGame {
  return { type: 'START_GAME', gameId }
}

export interface PlayerGuess extends Request {
  type: 'PLAYER_GUESS',
  clientGuess: ClientGuess
}

export function playerGuess(clientGuess: ClientGuess): PlayerGuess {
  return { type: 'PLAYER_GUESS', clientGuess }
}

export interface NextRound extends Request {
  type: 'NEXT_ROUND',
  gameId: string,
  userId: string
}

export function nextRound(gameId: string, userId: string): NextRound {
  return { type: 'NEXT_ROUND', gameId, userId: userId }
}

export interface ToggleString extends Request {
  type: 'TOGGLE_STRING',
  gameId: string,
  string: number
}

export function toggleString(gameId: string, string: number): ToggleString {
  return { type: 'TOGGLE_STRING', gameId, string }
}

export interface ToggleAccidental extends Request {
  type: 'TOGGLE_ACCIDENTAL',
  gameId: string,
  accidental: Accidental
}

export function toggleAccidental(gameId: string, accidental: Accidental): ToggleAccidental {
  return { type: 'TOGGLE_ACCIDENTAL', gameId, accidental };
}

export interface SetRoundCount extends Request {
  type: 'SET_ROUND_COUNT',
  gameId: string, 
  roundCount: number
}

export function setRoundCount(gameId: string, roundCount: number): SetRoundCount {
  return { type: 'SET_ROUND_COUNT', gameId, roundCount };
}

export interface StartCountdown extends Request {
  type: 'START_COUNTDOWN',
  gameId: string
}

export function startCountdown(gameId: string): StartCountdown {
  return { type: 'START_COUNTDOWN', gameId }
}

export interface StartRoundCountdown extends Request {
  type: 'START_ROUND_COUNTDOWN',
  gameId: string,
}

export function startRoundCountdown(gameId: string): StartRoundCountdown {
  return { type: 'START_ROUND_COUNTDOWN', gameId }
}
